codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -69,409 +71,1060 @@
+                    // ignore if child channel was already created.
+                    break;
+                }
+                // fall-trough
+                DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(stream);
+                ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+                if (future.isDone()) {
+                    registerDone(future);
+                } else {
+                    future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
+                }
+                break;
+            case CLOSED:
+                DefaultHttp2StreamChannel channel = ((Http2MultiplexCodecStream) stream).channel;
+                if (channel != null) {
+                    channel.streamClosed();
+                }
+                break;
+            default:
+                // ignore for now
+                break;
         }
     }
 
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
+    @Override
+    final void onHttp2StreamWritabilityChanged(ChannelHandlerContext ctx, Http2FrameStream stream, boolean writable) {
+        (((Http2MultiplexCodecStream) stream).channel).writabilityChanged(writable);
+    }
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
+    final Http2StreamChannel newOutboundStream() {
+        return new DefaultHttp2StreamChannel(newStream());
     }
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
-        } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
+    @Override
+    final void onHttp2FrameStreamException(ChannelHandlerContext ctx, Http2FrameStreamException cause) {
+        Http2FrameStream stream = cause.stream();
+        DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
+
+        try {
+            childChannel.pipeline().fireExceptionCaught(cause.getCause());
+        } finally {
+            childChannel.unsafe().closeForcibly();
         }
     }
 
-    void writeFromStreamChannel(Object msg, boolean flush) {
-        writeFromStreamChannel(msg, ctx.newPromise(), flush);
-    }
+    private void onHttp2StreamFrame(DefaultHttp2StreamChannel childChannel, Http2StreamFrame frame) {
+        switch (childChannel.fireChildRead(frame)) {
+            case READ_PROCESSED_BUT_STOP_READING:
+                if (childChannel.fireChildReadComplete()) {
+                    flushNeeded = true;
```

Comment:
https://github.com/netty/netty/pull/6896#discussion_r132304246

> can we just set `flushNeeded` in `fireChildReadComplete`? it looks like that is the only way it is used.

This is what I had in mind. This will help remove extra conditional checks on `fireChildReadComplete`.

```java
        void fireChildReadComplete() {
            assert eventLoop().inEventLoop();
            try {
                if (readInProgress) {
                    inFireChannelReadComplete = true;
                    readInProgress = false;
                    unsafe().recvBufAllocHandle().readComplete();
                    pipeline().fireChannelReadComplete();
                }
                flushNeeded |= flushPending;
            } finally {
                inFireChannelReadComplete = false;
                flushPending = false;
            }
        }
```